elastic/kibana
# [Reporting/Cloud] Soft disable reporting functionality on 1GB Kibanas

## Summary

Initial work for soft-disabling rather than crashing Kibana when generating reports on Kibana instances with low available RAM. This work ensures that Kibana server responds with the appropriate messages via our reporting HTTP APIs when generating a report.

## How to test

1. Start Kibana & ES
2. Enable a trial license
3. Either test on 1GB Cloud deployment (see this PR https://github.com/elastic/kibana/pull/130846) or, if testing locally replace `if (this.systemHasInsufficientMemory()) {` with `if (true) {` in `screenshotting/screenshots/index.ts` to trigger that code path.
4. Generate a PDF or PNG report
5. See the error message in the UI
6. See the error message in the report info panel flyout that contains a link

## Screenshots

<img width="458" alt="Screenshot 2022-04-22 at 12 27 14" src="https://user-images.githubusercontent.com/8155004/164694037-a34a22cc-ee2c-4378-800b-80abdac56cf0.png">
<img width="477" alt="Screenshot 2022-04-22 at 12 28 04" src="https://user-images.githubusercontent.com/8155004/164694052-ac508595-b064-4fc0-85b2-8ffb84a14670.png">




## Implementation details and notes

* The `InsufficientMemoryAvailableOnCloudError` will be thrown on Cloud when we are running on a Kibana with <2GB RAM. Specifically, the error is thrown when we call `getScreenshots` exposed by the screenshotting plugin. In this way we block all screenshots from running.
* Reporting now knows about this error type and sets an appropriate message explaining what is wrong. Additionally, the UI shows a message with a link, guiding users to how the can get around the issue
* For `ReportingError` classes it is convenient to read `this.code` and `ClassName.code` I've opted to make `code` both an instance and static property name
* In collaboration with Cloud stakeholders, we have decided that reading `os.totalmem` is sufficient for informing the soft disable decision
* No new methods were added to screenhotting's API, although we may have to consider this if we want to pre-emptively disable the "generate report" button

## Related to

https://github.com/elastic/kibana/pull/129989
 

### Checklist

Delete any items that are not applicable to this PR.


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Does not stop from generating reports on Cloud | Low | High | Must be tested manually on Cloud |


### For maintainers
